File: examples/java/customer/src/main/java/dev/angzarr/examples/customer/DefaultCustomerLogic.java
package dev.angzarr.examples.customer;

import com.google.protobuf.Any;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Timestamp;
import examples.Domains.CustomerCreated;
import examples.Domains.LoyaltyPointsAdded;
import examples.Domains.LoyaltyPointsRedeemed;
import dev.angzarr.EventBook;
import dev.angzarr.EventPage;
import dev.angzarr.Snapshot;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import static net.logstash.logback.argument.StructuredArguments.kv;

import java.time.Instant;
import java.util.List;

/**
 * Default implementation of customer business logic.
 */
public class DefaultCustomerLogic implements CustomerLogic {
    private static final Logger logger = LoggerFactory.getLogger(DefaultCustomerLogic.class);

    @Override
    public CustomerState rebuildState(EventBook eventBook) {
        CustomerState state = CustomerState.empty();

        if (eventBook == null || eventBook.getPagesList().isEmpty()) {
            return state;
        }

        // Start from snapshot if present
        if (eventBook.hasSnapshot() && eventBook.getSnapshot().hasState()) {
            try {
                var snapState = eventBook.getSnapshot().getState()
                    .unpack(examples.Domains.CustomerState.class);
                state = new CustomerState(
                    snapState.getName(),
                    snapState.getEmail(),
                    snapState.getLoyaltyPoints(),
                    snapState.getLifetimePoints()
                );
            } catch (InvalidProtocolBufferException e) {
                logger.warn("Failed to unpack snapshot state", e);
            }
        }

        // Apply events
        for (EventPage page : eventBook.getPagesList()) {
            if (!page.hasEvent()) {
                continue;
            }

            Any event = page.getEvent();
            state = applyEvent(state, event);
        }

        return state;
    }

    private CustomerState applyEvent(CustomerState state, Any event) {
        String typeUrl = event.getTypeUrl();

        try {
            if (typeUrl.endsWith("CustomerCreated")) {
                CustomerCreated created = event.unpack(CustomerCreated.class);
                return new CustomerState(created.getName(), created.getEmail(), 0, 0);
            } else if (typeUrl.endsWith("LoyaltyPointsAdded")) {
                LoyaltyPointsAdded added = event.unpack(LoyaltyPointsAdded.class);
                return state.withLoyaltyPoints(added.getNewBalance())
                    .addLifetimePoints(added.getPoints());
            } else if (typeUrl.endsWith("LoyaltyPointsRedeemed")) {
                LoyaltyPointsRedeemed redeemed = event.unpack(LoyaltyPointsRedeemed.class);
                return state.withLoyaltyPoints(redeemed.getNewBalance());
            }
        } catch (InvalidProtocolBufferException e) {
            logger.warn("Failed to unpack event: {}", typeUrl, e);
        }

        return state;
    }

    @Override
    public EventBook handleCreateCustomer(CustomerState state, String name, String email)
            throws CommandValidationException {
        if (state.exists()) {
            throw CommandValidationException.failedPrecondition("Customer already exists");
        }

        if (name == null || name.isBlank()) {
            throw CommandValidationException.invalidArgument("Customer name is required");
        }

        if (email == null || email.isBlank()) {
            throw CommandValidationException.invalidArgument("Customer email is required");
        }

        logger.info("creating_customer", kv("name", name), kv("email", email));

        CustomerCreated event = CustomerCreated.newBuilder()
            .setName(name)
            .setEmail(email)
            .setCreatedAt(nowTimestamp())
            .build();

        return createEventBook(event);
    }

    @Override
    public EventBook handleAddLoyaltyPoints(CustomerState state, int points, String reason)
            throws CommandValidationException {
        if (!state.exists()) {
            throw CommandValidationException.failedPrecondition("Customer does not exist");
        }

        if (points <= 0) {
            throw CommandValidationException.invalidArgument("Points must be positive");
        }

        int newBalance = state.loyaltyPoints() + points;

        logger.info("adding_loyalty_points",
            kv("points", points), kv("new_balance", newBalance), kv("reason", reason));

        LoyaltyPointsAdded event = LoyaltyPointsAdded.newBuilder()
            .setPoints(points)
            .setNewBalance(newBalance)
            .setReason(reason != null ? reason : "")
            .build();

        return createEventBook(event);
    }

    @Override
    public EventBook handleRedeemLoyaltyPoints(CustomerState state, int points, String redemptionType)
            throws CommandValidationException {
        if (!state.exists()) {
            throw CommandValidationException.failedPrecondition("Customer does not exist");
        }

        if (points <= 0) {
            throw CommandValidationException.invalidArgument("Points must be positive");
        }

        if (points > state.loyaltyPoints()) {
            throw CommandValidationException.failedPrecondition(
                String.format("Insufficient points: have %d, need %d",
                    state.loyaltyPoints(), points));
        }

        int newBalance = state.loyaltyPoints() - points;

        logger.info("redeeming_loyalty_points",
            kv("points", points), kv("new_balance", newBalance), kv("redemption_type", redemptionType));

        LoyaltyPointsRedeemed event = LoyaltyPointsRedeemed.newBuilder()
            .setPoints(points)
            .setNewBalance(newBalance)
            .setRedemptionType(redemptionType != null ? redemptionType : "")
            .build();

        return createEventBook(event);
    }

    private EventBook createEventBook(com.google.protobuf.Message event) {
        EventPage page = EventPage.newBuilder()
            .setNum(0)
            .setEvent(Any.pack(event))
            .setCreatedAt(nowTimestamp())
            .build();

        return EventBook.newBuilder()
            .addPages(page)
            .build();
    }

    private Timestamp nowTimestamp() {
        Instant now = Instant.now();
        return Timestamp.newBuilder()
            .setSeconds(now.getEpochSecond())
            .setNanos(now.getNano())
            .build();
    }
}

File: examples/kotlin/customer/src/main/kotlin/dev/angzarr/examples/customer/DefaultCustomerLogic.kt
package dev.angzarr.examples.customer

import dev.angzarr.EventBook
import dev.angzarr.EventPage
import com.google.protobuf.Any
import com.google.protobuf.Timestamp
import examples.Domains.CustomerCreated
import examples.Domains.LoyaltyPointsAdded
import examples.Domains.LoyaltyPointsRedeemed
import org.slf4j.LoggerFactory

/**
 * Default implementation of customer business logic.
 */
class DefaultCustomerLogic : CustomerLogic {

    private val logger = LoggerFactory.getLogger(DefaultCustomerLogic::class.java)

    override fun rebuildState(eventBook: EventBook?): CustomerState {
        if (eventBook == null || eventBook.pagesList.isEmpty()) {
            return CustomerState.empty()
        }

        var state = CustomerState.empty()

        // Start from snapshot if present
        eventBook.snapshot?.state?.let { snapAny ->
            if (snapAny.`is`(examples.Domains.CustomerState::class.java)) {
                val snapState = snapAny.unpack(examples.Domains.CustomerState::class.java)
                state = CustomerState(
                    name = snapState.name,
                    email = snapState.email,
                    loyaltyPoints = snapState.loyaltyPoints,
                    lifetimePoints = snapState.lifetimePoints
                )
            }
        }

        // Apply events
        for (page in eventBook.pagesList) {
            val event = page.event ?: continue
            state = applyEvent(state, event)
        }

        return state
    }

    private fun applyEvent(state: CustomerState, event: Any): CustomerState {
        return when {
            event.`is`(CustomerCreated::class.java) -> {
                val e = event.unpack(CustomerCreated::class.java)
                state.copy(name = e.name, email = e.email)
            }
            event.`is`(LoyaltyPointsAdded::class.java) -> {
                val e = event.unpack(LoyaltyPointsAdded::class.java)
                state.copy(
                    loyaltyPoints = e.newBalance,
                    lifetimePoints = state.lifetimePoints + e.points
                )
            }
            event.`is`(LoyaltyPointsRedeemed::class.java) -> {
                val e = event.unpack(LoyaltyPointsRedeemed::class.java)
                state.copy(loyaltyPoints = e.newBalance)
            }
            else -> state
        }
    }

    override fun handleCreateCustomer(state: CustomerState, name: String, email: String): EventBook {
        if (state.exists()) {
            throw CommandValidationException.failedPrecondition("Customer already exists")
        }

        if (name.isEmpty()) {
            throw CommandValidationException.invalidArgument("Customer name is required")
        }
        if (email.isEmpty()) {
            throw CommandValidationException.invalidArgument("Customer email is required")
        }

        logger.info("creating_customer name={} email={}", name, email)

        val event = CustomerCreated.newBuilder()
            .setName(name)
            .setEmail(email)
            .setCreatedAt(nowTimestamp())
            .build()

        return createEventBook(event)
    }

    override fun handleAddLoyaltyPoints(state: CustomerState, points: Int, reason: String): EventBook {
        if (!state.exists()) {
            throw CommandValidationException.failedPrecondition("Customer does not exist")
        }

        if (points <= 0) {
            throw CommandValidationException.invalidArgument("Points must be positive")
        }

        val newBalance = state.loyaltyPoints + points

        logger.info("adding_loyalty_points points={} new_balance={} reason={}", points, newBalance, reason)

        val event = LoyaltyPointsAdded.newBuilder()
            .setPoints(points)
            .setNewBalance(newBalance)
            .setReason(reason)
            .build()

        return createEventBook(event)
    }

    override fun handleRedeemLoyaltyPoints(state: CustomerState, points: Int, redemptionType: String): EventBook {
        if (!state.exists()) {
            throw CommandValidationException.failedPrecondition("Customer does not exist")
        }

        if (points <= 0) {
            throw CommandValidationException.invalidArgument("Points must be positive")
        }
        if (points > state.loyaltyPoints) {
            throw CommandValidationException.failedPrecondition(
                "Insufficient points: have ${state.loyaltyPoints}, need $points"
            )
        }

        val newBalance = state.loyaltyPoints - points

        logger.info("redeeming_loyalty_points points={} new_balance={} redemption_type={}", points, newBalance, redemptionType)

        val event = LoyaltyPointsRedeemed.newBuilder()
            .setPoints(points)
            .setNewBalance(newBalance)
            .setRedemptionType(redemptionType)
            .build()

        return createEventBook(event)
    }

    private fun createEventBook(event: com.google.protobuf.Message): EventBook {
        val page = EventPage.newBuilder()
            .setNum(0)
            .setEvent(Any.pack(event))
            .setCreatedAt(nowTimestamp())
            .build()

        return EventBook.newBuilder()
            .addPages(page)
            .build()
    }

    private fun nowTimestamp(): Timestamp = Timestamp.newBuilder()
        .setSeconds(System.currentTimeMillis() / 1000)
        .build()
}
